mjpc/planners/mppi/planner.cc

```diff
@@ -0,0 +1,508 @@
+  ncandidates = std::min(ncandidates, num_trajectory);  // unused
+  ResizeMjData(model, pool.NumThreads());
+
+  // ----- rollout noisy policies ----- //
+  // start timer
+  auto rollouts_start = std::chrono::steady_clock::now();
+
+  // simulate noisy policies
+  this->Rollouts(num_trajectory, horizon, pool);
+
+  // stop timer
+  rollouts_compute_time = GetDuration(rollouts_start);
+
+  return ncandidates;
+}
+
+// optimize nominal policy using random sampling
+void MPPIPlanner::OptimizePolicy(int horizon, ThreadPool& pool) {
+  // resample nominal policy to current time
+  OptimizePolicyCandidates(1, horizon, pool);  // executes noisy rollouts
+  this->UpdateNominalPolicy(horizon);
+
+  // ----- update policy ----- //
+  // start timer
+  auto policy_update_start = std::chrono::steady_clock::now();
+
+  CopyCandidateToPolicy(0);
+
+  // improvement: compare nominal to winner
+  // TODO(ahl): replace the functionality of the winner here
+  // double best_return = trajectory[0].total_return;
+  // improvement = mju_max(best_return - trajectory[winner].total_return, 0.0);
+  improvement = 0.0;
+
+  // stop timer
+  policy_update_compute_time = GetDuration(policy_update_start);
+}
+
+// compute trajectory using nominal policy
+void MPPIPlanner::NominalTrajectory(int horizon, ThreadPool& pool) {
+  // set policy
+  auto nominal_policy = [&cp = candidate_policy[0]](
+                            double* action, const double* state, double time) {
+    cp.Action(action, state, time);
+  };
+
+  // rollout nominal policy
+  trajectory[0].Rollout(nominal_policy, task, model, data_[0].get(),
+                        state.data(), time, mocap.data(), userdata.data(),
+                        horizon);
+}
+
+// set action from policy
+void MPPIPlanner::ActionFromPolicy(double* action, const double* state,
+                                   double time, bool use_previous) {
+  const std::shared_lock<std::shared_mutex> lock(mtx_);
+  if (use_previous) {
+    previous_policy.Action(action, state, time);
+  } else {
+    policy.Action(action, state, time);
+  }
+}
+
+// update policy via resampling
+void MPPIPlanner::UpdateNominalPolicy(int horizon) {
+  // dimensions
+  int num_trajectory = num_trajectory_;  // in case this gets updated in GUI
+  int num_spline_points = candidate_policy[0].num_spline_points;
+
+  // set time
+  double nominal_time = time;
+  double time_shift = mju_max(
+      (horizon - 1) * model->opt.timestep / (num_spline_points - 1), 1.0e-5);
+
+  // get spline points
+  for (int t = 0; t < num_spline_points; t++) {
+    times_scratch[t] = nominal_time;
+    nominal_time += time_shift;
+  }
+
+  // compute MPPI update terms
+  double weight = 0.0;  // storage for intermediate weights
+  double denom = 0.0;
+  std::vector<double> numer(policy.num_parameters, 0.0);
+
+  for (int i = 0; i < num_trajectory; i++) {
+    // update denominator
+    weight = std::exp(-trajectory[i].total_return / lambda);
```

Comment:
This should be `weight = std::exp(-(trajectory[i].total_return - min_total_return) / lambda);`.

In my python code it breaks if I don't have that.

However, I just tried that on my machine and it doesn't fix the problem

Reply:
ah thanks, forgot about the baseline. i'll fix it (this also means i should add back in the sorting functionality that i removed + the winner)